File: Patterns/Iterator Pattern/CSharp/Iterator/PancakeHouseMenuIterator.cs
using System.Collections.Generic;

using MenuItem = Menu.MenuItem;

namespace Iterator
{
    public class PancakeHouseMenuIterator : Iterator
    {
        private List<MenuItem> items;
        private int position;

        public PancakeHouseMenuIterator(List<MenuItem> items) 
        {
            this.items = items;
            position = 0;
        }

        public MenuItem Next() => items[position++];

        public bool HasNext() 
        {
            if (position >= items.Count || items[position] == null) return false;
            return true;
        }
    }
}

File: Patterns/Iterator Pattern/Cpp/Iterator/PancakeHouseMenuIterator.cpp
#pragma once

#include <vector>
#include "Iterator.cpp"
#include "../Menu/MenuItem.cpp"

using namespace std;

class PancakeHouseMenuIterator : public Iterator {
private:
    vector<MenuItem*> items;
    int position;

public:
    PancakeHouseMenuIterator(vector<MenuItem*> items) : items(items) {
        position = 0;
    }

    MenuItem* next() override {
        return items[position++];
    }

    bool hasNext() override {
        if (position >= items.size() || items[position] == nullptr) return false;
        return true;
    }
};